Disable GSP Firmware for select instance families.

larvacea opened the pull request:
**Description of changes:**

For some instance families in AWS, we wish to disable GSP firmware download in the NVIDIA kmod. We can do this by creating a configuration file in /etc/modprobe.d with the desired option, conditionally based on the instance family we fetch from IMDS. This must happen before we invoke driverdog to load the kmod. In bottlerocket, /etc is ephemeral, so we have to do this on each boot.

**Testing done:**

Manual testing demonstrates that the setting takes effect on the desired instance families, and does not take effect on other instance families. Tested on each kernel version, on aws-eks and aws-ecs variants.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

larvacea committed:
Disable GSP Firmware for select instance families.

larvacea committed:
Run rustfmt on shibaken: oops.

bcressey commented on packages/kmod-5.10-nvidia/disable-gsp.service.in:
```diff
@@ -0,0 +1,23 @@
+Type=oneshot
+ExecCondition=/usr/bin/shibaken does-value-start-with instance-type \
+  --value g4dn. --value g5. --value g5g.
+ExecStart=-/usr/bin/cp \
+  PREFIX/share/factory/etc/modprobe.d/disable-gsp.conf \
+  /etc/modprobe.d/disable-gsp.conf
```
Why ignore the exit code from `cp`?

larvacea replied:
My reasoning:

- If we do not ignore the exit code and cp fails, that cascades into later units and may prevent boot.
- If we do ignore the exit code and cp fails, we boot as we would today. We have not disabled GSP firmware download, but the kmods load and boot proceeds.

I think that "boot proceeds" is the wrong outcome. In the case of intermittent failures in the cp command, having boot proceed but not do the expected "disable GSP firmware download" leaves the customer with an instance that does not work as expected. Better to have the dead instance, since orchestrators know how to deal with that scenario.

bcressey commented on packages/kmod-5.15-nvidia/disable-gsp.service.in:
```diff
@@ -0,0 +1,23 @@
+[Unit]
+Description=Disable GSP for selected instance types
+RequiresMountsFor=/etc/modprobe.d PREFIX/share/factory/etc/modprobe.d
```
Only `/etc` is a mount that we would nominally be concerned about.

bcressey commented on packages/kmod-5.15-nvidia/disable-gsp.service.in:
```diff
@@ -0,0 +1,23 @@
+RefuseManualStop=true
+
+[Service]
+Type=oneshot
+ExecCondition=/usr/bin/shibaken does-value-start-with instance-type \
```
One drawback is that because we transition targets several times (to configured, then to multi-user) this check will be repeated on other instance types (since the service won't have started and won't still be running).

You can avoid that by creating and checking for a sentinel file:
```
[Unit]
...
ConditionPathExists=!/etc/.gsp-disable-check-ran

[Service]
Type=oneshot
# This always returns true, but creates the sentinel file to prevent future checks.
ExecCondition=/usr/bin/touch /etc/.gsp-disable-check-ran
# Only proceed on specific instance types.
ExecCondition=/usr/bin/shibaken does-value-start-with instance-type \
  --value g4dn. --value g5. --value g5g.
...
```

larvacea replied:
I can do that. The actual state changing cp execution is idempotent, but the IMDS query is not free, so let's not repeat it.

bcressey commented on packages/kmod-5.15-nvidia/disable-gsp.service.in:
```diff
@@ -0,0 +1,23 @@
+[Service]
+Type=oneshot
+ExecCondition=/usr/bin/shibaken does-value-start-with instance-type \
+  --value g4dn. --value g5. --value g5g.
```
Is this needed on `g6` also?

larvacea replied:
Probably. I'm not able to launch a g6 instance to see what nvidia_smi would tell me.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -89,6 +97,15 @@ sed \
+ install -p -m 0644 \
+   disable-gsp.service \
+   %{buildroot}%{_cross_unitdir}
+install -d %{buildroot}%{_cross_factorydir}/etc/modprobe.d
```
This is OK, but putting it in `/usr/share/factory` invites the misconception that `tmpfiles` is supposed to manage it. It could just as easily go in `%{_cross_datadir}/nvidia/modprobe.d` and avoid that confusion.

bcressey commented on packages/kmod-5.15-nvidia/disable-gsp-tmpfiles.conf:
```diff
@@ -0,0 +1 @@
+d /etc/modprobe.d 0750 root root -
```
This should either go in `release.spec`, or we should change the unit to `mkdir /etc/modprobe.d` instead. Otherwise it's weird for a kmod package to own creating a system directory.

larvacea replied:
Changed to use systemd-tmpfiles.

bcressey commented on sources/api/shibaken/src/value_equal.rs:
```diff
@@ -0,0 +1,59 @@
+    value: Vec<String>,
+}
+
+impl ValueStartsWith {
+    pub(crate) async fn run(self) -> Result<ExitCode> {
+        if self.variable_name != "instance-type" {
+            log::info!(
+                "Unknown variable name {}, returning false.",
+                self.variable_name
+            );
+            return Ok(ExitCode::FAILURE);
+        }
+        if self.value.is_empty() {
```
Would be good to add unit tests for this, if we can mock out the IMDS client.

bcressey commented:
> For some instance families in AWS, we wish to disable GSP firmware download in the NVIDIA kmod.

Are we able to do this by checking for the hardware in question, or querying for some shared attribute with `nvidia-smi`, rather than checking a list of instance types?

The reason I ask is that hardware should be the same across a set of instance types, while code that checks the instance type will need to be revisited if newer g4 or g5 types are launched.

larvacea commented:
I can take a look. As I understand the technical note today, we must set the option before we load the nvidia kmod. I don't know if nvidia_smi would run without that kmod loaded, and I don't know if there's some other way to query the GPU and discover whether it should or should not download GSP firmware. So let me go looking. I agree that making the decision based on what hardware we see would be preferable to pattern-matching the EC2 instance family.

larvacea committed:
PR comment implementation

Create sentinel file. Store modprobe file in datadir rather than
factorydir. Fail the service start if cp fails.

larvacea commented:
Updated branch to incorporate @bcressey's suggestions. I will rebase and squash before merging, eventually, but kept a separate commit to make it fractionally easier to see what's changed, when.

I have not added to the list of instance families here since I have not been able to launch on the newer instance types.
